trinity/protocol/eth/managers.py

```diff
@@ -0,0 +1,57 @@
+from typing import (
+    Tuple,
+    Type,
+    TYPE_CHECKING,
+)
+
+from eth_typing import BlockIdentifier
+
+from eth.rlp.headers import BlockHeader
+
+from p2p.protocol import (
+    Command,
+)
+
+from trinity.protocol.common.managers import (
+    BaseRequestManager as _BaseRequestManager,
+)
+
+from .commands import BlockHeaders
+from .requests import HeaderRequest
+
+if TYPE_CHECKING:
+    from .peer import ETHPeer  # noqa: F401
+
+
+BaseRequestManager = _BaseRequestManager[
+    'ETHPeer',
+    HeaderRequest,
+    Tuple[BlockHeader, ...],
+    Tuple[BlockHeader, ...],
+]
+
+
+class GetBlockHeadersRequestManager(BaseRequestManager):
+    msg_queue_maxsize = 100
+
+    _response_msg_type: Type[Command] = BlockHeaders
+
+    async def __call__(self,  # type: ignore
```

Comment:
The signature doesn't align with the parent. The signature here (the child) is

```py
 __call__(self, # type: ignore
                       block_number_or_hash: BlockIdentifier,
                       max_headers: int = None,
                       skip: int = 0,
                       reverse: bool = True,
                       timeout: int = None) -> Tuple[BlockHeader, ...]
```

The signature in the parent is:

```py
def __call__(self) -> ReturnType:
```

While the `ReturnType` matches just fine, notice that the parent method is defined as one without any arguments whereas the child implementation does have arguments. That is violating the type safety rules.

What we could do is to define `__call__` in the parent as `__call__(self, req_params: TRequest) -> TReturn:`
Then, in the child it's perfectly valid to implement it as `__call__(self, req_params: HeaderRequest) -> Tuple[BlockHeader, ...]:`. So, instead of passing the arguments as individual arguments that mypy doesn't know about, we use the type of the request that the subclass is specialized for (as specified via `BaseRequestManager[T1, T2, T3, T4]`

Hope that helps.

Reply:
Btw, there is also the alternative to define `__call__` in parent and child as `*args: Any` but that gives up type safety whereas the previous solution doesn't.

Reply:
This is intentional.  I spent a decent amount of time trying to figure out a sane way to do this and this is the result.  Here are the properties I was able to achieve.

- the `__call__` method is required to be implemented on subclasses.
- call sites which call this class still do correct type checks on the function signature (i.e. we don't loose type safety when this function is being called)
- all `__call__` implementations on subclasses require a `# type: ignore` comment.
- we don't have to manually instantiate a `Request` object in order to use the API (**this one is important**)

I've added a comment to both of the `Manager` subclasses noting this.

I'm open to alternate solutions but all that I've seen have trade-offs that I'm not ok with (like requiring a `Request` object to use the API.

Reply:
It's not a big deal, but it exposes the underlying implementation which bothers me.  It means every single call site for `foo` must import `Request` when really the `Request` class is an internal API that callers should not be exposed to.

Reply:
I see. It doesn't *have* to be `Request` though. We could add one more generic type which could be set to something as lightweight as `Tuple[int, int, str]` and make call sites look like `foo((1, 2, 'bar'))`.

I wanted to find out if there's any other way how we could constrain an abstract method on only the return type and created a question on SO. But there doesn't seem to be anything and someone also [pointed out](https://stackoverflow.com/questions/51720280/is-it-possible-to-constrain-an-abstract-base-method-only-on-the-return-type-bu/51724793#comment90401496_51720280) that it's violating the [Liskov Substitution Principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle).

But anyway, I didn't mean to bikeshed about it. If it works for you, it works for me :)

Reply:
I don't think the `Tuple[...]` example works because the signatures of these will vary widely in both argument type and count, so we'd have to use `Tuple[Any, ...]`

With this we'd end up with satisfying the type checker while losing a degree of type safety on the actual function calls (at least I think we would).

Right now this works for me.  I'm open to an alternative if someone wants to take the time to figure it out.